### PR TITLE
CreateExperiment API.

### PR DESCRIPTION
**Pull request by msvinaykumar**
Signed-off-by: msvinaykumar <vinakuma@redhat.com>

### CODE REVIEW

**Comment by chandrams** [comment]
@msvinaykumar - Have you tested this? I get the below error, while creating the experiment. Am I missing anything here?
```
java.lang.NumberFormatException: For input string: "15min"
	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(Unknown Source)
	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(Unknown Source)
	at java.base/java.lang.Double.parseDouble(Unknown Source)
	at com.google.gson.stream.JsonReader.nextDouble(JsonReader.java:909)
	at com.google.gson.Gson$1.read(Gson.java:362)
	at com.google.gson.Gson$1.read(Gson.java:356)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:130)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:221)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:130)
	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:221)
	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
	at com.google.gson.internal.bind.ArrayTypeAdapter.read(ArrayTypeAdapter.java:72)
	at com.google.gson.Gson.fromJson(Gson.java:991)
	at com.google.gson.Gson.fromJson(Gson.java:956)
	at com.google.gson.Gson.fromJson(Gson.java:905)
	at com.google.gson.Gson.fromJson(Gson.java:876)
	at com.autotune.analyzer.services.CreateExperiment.doPost(CreateExperiment.java:101)

```

**Comment by msvinaykumar** [comment]
I was think

> @msvinaykumar - Have you tested this? I get the below error, while creating the experiment. Am I missing anything here?
> 
> ```
> java.lang.NumberFormatException: For input string: "15min"
> 	at java.base/jdk.internal.math.FloatingDecimal.readJavaFormatString(Unknown Source)
> 	at java.base/jdk.internal.math.FloatingDecimal.parseDouble(Unknown Source)
> 	at java.base/java.lang.Double.parseDouble(Unknown Source)
> 	at com.google.gson.stream.JsonReader.nextDouble(JsonReader.java:909)
> 	at com.google.gson.Gson$1.read(Gson.java:362)
> 	at com.google.gson.Gson$1.read(Gson.java:356)
> 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:130)
> 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:221)
> 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$1.read(ReflectiveTypeAdapterFactory.java:130)
> 	at com.google.gson.internal.bind.ReflectiveTypeAdapterFactory$Adapter.read(ReflectiveTypeAdapterFactory.java:221)
> 	at com.google.gson.internal.bind.TypeAdapterRuntimeTypeWrapper.read(TypeAdapterRuntimeTypeWrapper.java:41)
> 	at com.google.gson.internal.bind.ArrayTypeAdapter.read(ArrayTypeAdapter.java:72)
> 	at com.google.gson.Gson.fromJson(Gson.java:991)
> 	at com.google.gson.Gson.fromJson(Gson.java:956)
> 	at com.google.gson.Gson.fromJson(Gson.java:905)
> 	at com.google.gson.Gson.fromJson(Gson.java:876)
> 	at com.autotune.analyzer.services.CreateExperiment.doPost(CreateExperiment.java:101)
> ```

Hello @chandrams Thank you for looking into this

I was thinking to have format to represent time like this instead of combined string.

     "trial_settings": {
            "measurement_duration": "15",
            "format": "M"
        },
Where "M" stands for minutes  "H" for hours and so on.

Can you please suggest if you are ok with this , Else will change it back to combined string like 15Min.
And I will also try to find if any industry standard format to represent time in JSON. Above one I just copied from Kube apis.

**Comment by chandrams** [comment]
I think we should look at the ease of converting the specified format in json using java date/time methods into whatever format we require. I see in ISO_8601 they have this format 'PYMDHMS'. Can you check if we have date/time methods in java to convert this.

File: src/main/java/com/autotune/common/k8sObjects/KruizeObject.java [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+        return experimentUseCaseType;
+    }
+
+    public void setExperimentUseCaseType(ExperimentUseCaseType experimentUseCaseType) {
+        this.experimentUseCaseType = experimentUseCaseType;
+    }
+
+    @Override
+    public String toString() {
+        return "AutotuneObject{" +
```
Comment (chandrams) [inline src/main/java/com/autotune/common/k8sObjects/KruizeObject.java:206]: Should this be KruizeObject?

Comment (msvinaykumar) [reply src/main/java/com/autotune/common/k8sObjects/KruizeObject.java]: yes .done

**Comment by chandrams** [comment]
@msvinaykumar - Also update the API doc with the changed measurement duration format

**Comment by msvinaykumar** [comment]
> 

Right now im going ahead with String "15min" as per initial draft, Seems we need separate pr to take up formats across all Modules which receives date and time.

**Comment by chandrams** [comment]
@msvinaykumar - Can you please check updateResults, I get the below exception when I tested with the changes in this PR. 

```
2022-12-2807:52:02.378 ERROR [qtp551016187-31][updateResults.java(82)]-Exception due to :Cannot invoke "com.google.gson.JsonObject.get(String)" because "exp" is null
2022-12-2807:52:02.378 ERROR [qtp551016187-31][updateResults.java(104)]-java.lang.NullPointerException: Cannot invoke "com.google.gson.JsonObject.get(String)" because "exp" is null
java.lang.NullPointerException: Cannot invoke "com.google.gson.JsonObject.get(String)" because "exp" is null
	at com.autotune.analyzer.services.updateResults.doPost(updateResults.java:70)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:707)
	at javax.servlet.http.HttpServlet.service(HttpServlet.java:790)
	at org.eclipse.jetty.servlet.ServletHolder.handle(ServletHolder.java:799)
	at org.eclipse.jetty.servlet.ServletHandler.doHandle(ServletHandler.java:550)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextHandle(ScopedHandler.java:233)
	at org.eclipse.jetty.server.handler.ContextHandler.doHandle(ContextHandler.java:1434)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:188)
	at org.eclipse.jetty.servlet.ServletHandler.doScope(ServletHandler.java:501)
	at org.eclipse.jetty.server.handler.ScopedHandler.nextScope(ScopedHandler.java:186)
	at org.eclipse.jetty.server.handler.ContextHandler.doScope(ContextHandler.java:1349)
	at org.eclipse.jetty.server.handler.ScopedHandler.handle(ScopedHandler.java:141)
	at org.eclipse.jetty.server.handler.HandlerWrapper.handle(HandlerWrapper.java:127)
	at org.eclipse.jetty.server.Server.handle(Server.java:516)
	at org.eclipse.jetty.server.HttpChannel.lambda$handle$1(HttpChannel.java:400)
	at org.eclipse.jetty.server.HttpChannel.dispatch(HttpChannel.java:645)
	at org.eclipse.jetty.server.HttpChannel.handle(HttpChannel.java:392)
	at org.eclipse.jetty.server.HttpConnection.onFillable(HttpConnection.java:277)
	at org.eclipse.jetty.io.AbstractConnection$ReadCallback.succeeded(AbstractConnection.java:311)
	at org.eclipse.jetty.io.FillInterest.fillable(FillInterest.java:105)
	at org.eclipse.jetty.io.ChannelEndPoint$1.run(ChannelEndPoint.java:104)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.runTask(EatWhatYouKill.java:338)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.doProduce(EatWhatYouKill.java:315)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.tryProduce(EatWhatYouKill.java:173)
	at org.eclipse.jetty.util.thread.strategy.EatWhatYouKill.run(EatWhatYouKill.java:131)
	at org.eclipse.jetty.util.thread.ReservedThreadExecutor$ReservedThread.run(ReservedThreadExecutor.java:409)
	at org.eclipse.jetty.util.thread.QueuedThreadPool.runJob(QueuedThreadPool.java:883)
	at org.eclipse.jetty.util.thread.QueuedThreadPool$Runner.run(QueuedThreadPool.java:1034)
	at java.base/java.lang.Thread.run(Unknown Source)
```

**Comment by msvinaykumar** [comment]
thanks @chandrams  for testing this, I fixed this and should be fine now.

File: src/main/java/com/autotune/analyzer/data/ExperimentInterfaceImpl.java [hunk 2]
```diff
@@ -0,0 +1,55 @@
...
+
+    @Override
+    public boolean addExperiments(Map<String, KruizeObject> mainKruizeExperimentMap, List<KruizeObject> kruizeExperimentList) {
+        kruizeExperimentList.forEach(
+                (ao) -> {
+                    ao.setStatus(AnalyzerConstants.ExpStatus.QUEUED);
+                    ao.setExperimentId(Utils.generateID(toString()));
+                    mainKruizeExperimentMap.put(
+                            ao.getExperimentName(),
+                            ao
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/data/ExperimentInterfaceImpl.java:39]: `ao` is short for `autotuneObject`? Suggest you expand it and call it `kruizeObject`

File: src/main/java/com/autotune/analyzer/services/updateResults.java [hunk 3]
```diff
@@ -49,32 +50,45 @@
...
+
+
+    public static void main(String[] args) throws InvalidValueException {
+        Map<String, KruizeObject> testMap = new ConcurrentHashMap<>();
+        testMap.put("abc", new KruizeObject());
+
+        KruizeObject tmp = testMap.get(new String("abc1"));
+        System.out.println(tmp);
+
+    }
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/services/updateResults.java:63]: Is this some temp code? Why is this needed?

File: src/main/java/com/autotune/analyzer/utils/ExperimentUseCaseType.java [hunk 4]
```diff
@@ -0,0 +1,87 @@
...
+            } else {
+                throw new Exception("Invalid Mode " + ao.getMode() + " for target cluster as Remote.");
+            }
+        } else if (ao.getTargetCluster().equalsIgnoreCase(AnalyzerConstants.LOCAL)) {
+            if (ao.getMode().equalsIgnoreCase(AnalyzerConstants.MONITOR)) {
+                setLocalMonitoring(true);
+            } else if (ao.getMode().equalsIgnoreCase(AnalyzerConstants.EXPERIMENT)) {
+                setLocalExperiment(true);
+            } else {
+                throw new Exception("Invalid Mode " + ao.getMode() + " for target cluster as Local.");
```
Comment (dinogun) [inline src/main/java/com/autotune/analyzer/utils/ExperimentUseCaseType.java:48]: `ao` needs to be `kruizeObject`

File: src/main/java/com/autotune/common/data/GeneralDataHolder.java [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package com.autotune.common.data;
+
+/**
+ * General data carrier between functions.
+ */
+public class GeneralDataHolder {
```
Comment (dinogun) [inline src/main/java/com/autotune/common/data/GeneralDataHolder.java:21]: Needs a better/more descriptive name :)

File: src/main/java/com/autotune/common/k8sObjects/TrialSettings.java [hunk 6]
```diff
@@ -0,0 +1,39 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *******************************************************************************/
+package com.autotune.common.k8sObjects;
+
+import com.google.gson.annotations.SerializedName;
+
+public class TrialSettings {
```
Comment (dinogun) [inline src/main/java/com/autotune/common/k8sObjects/TrialSettings.java:20]: Trial settings needs to have all of the following fields as per this JSON
https://github.com/kruize/autotune/blob/mvp_demo/examples/em-json/GeneralPerfExperimentsTemplate.json
```
       "trial_settings": {
          "measurement_cycles": "3",
          "warmup_duration": "1min",
          "warmup_cycles": "3",
          "measurement_duration": "1min",
          "iterations": "3"
```

File: src/main/java/com/autotune/utils/AnalyzerConstants.java [hunk 7]
```diff
@@ -15,272 +15,310 @@
...
+        MONITORING,
+        EXPERIMENT
+    }
+
+    public enum TargetType {
+        LOCAL,
+        REMOTE
+    }
+
+    public enum ExpStatus {
```
Comment (dinogun) [inline src/main/java/com/autotune/utils/AnalyzerConstants.java:89]: Suggest to expand the name to `ExperimentStatus`

**Comment by msvinaykumar** [comment]
@dinogun  thank you for your review comments. And I have incorporated all of them.

**Review by dinogun — APPROVED** [review]
LGTM
